draft-ietf-quic-tls.md

```diff
@@ -760,14 +756,9 @@ and ignoring any outstanding Initial packets.
 
 ### Discarding Handshake Keys
 
-An endpoint MUST NOT discard its handshake keys until the TLS handshake is
-confirmed ({{handshake-confirmed}}).  An endpoint SHOULD discard its handshake
-keys as soon as it has confirmed the handshake.  Most application protocols
-will send data after the handshake, resulting in acknowledgements that allow
-both endpoints to discard their handshake keys promptly.  Endpoints that do
-not have reason to send immediately after completing the handshake MAY send
-ack-eliciting frames, such as PING, which will cause the handshake to be
-confirmed when they are acknowledged.
+An endpoint MUST discard its handshake keys when the TLS handshake is confirmed
+({{handshake-confirmed}}).  The server MUST send a HANDSHAKE_DONE frame within
+one round-trip time of handshake completion.
```

Comment:
I think "within one round-trip time" is too long. As this happens before both endpoints have good estimation of RTT, it could well be the case that the server's estimated RTT be much greater than the PTO of the client. In such case, it would seem to the client as if the CF had been lost.

I think we should say "SHOULD immediately send", with possibly acknowledging the fact that the server "can bundle the HANDSHAKE_DONE frame with 0.5 RTT data."

Reply:
I tend to agree with @kazuho point.  How about "The server MUST send HANDSHAKE_DONE with the first outgoing 1-RTT packet or within one round-trip time of handshake completion if no 1-RTT packets are sent."

@kazuho If the server is sending HANDSHAKE_DONE, isn't it no longer 0.5RTT data?

Reply:
I agree with @ianswett. The handshake isn't complete when we have 0.5RTT data. In fact, if the frame goes out then, we can easily have the client discard the handshake keys before getting ACK of its finished.

Reply:
@ianswett 
> If the server is sending HANDSHAKE_DONE, isn't it no longer 0.5RTT data?

Yes. This is 1 RTT, *not* 0.5 RTT.

Regarding when HANDSHAKE_DONE should be sent, my view is that it HANDSHAKE_DONE is essentially an ACK for the Handshake packet. Based on that view, I do not see a reason why it should be sent at a different timing than other acks for Handshake packets (i.e. send immediately, though an endpoint might wait for a short amount of time to bundle ACK with some other data).

In practice, there is no reason to delay the emission of HANDSHAKE_DONE as much as 1 RTT in HTTP/3. The server should be sending HANDSHAKE_DONE at an early moment for letting the client know that the handshake is done before the client retransmits CF, doing so has a higher probability of having CWND increased at an earlier point.